<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Reed Sorenson | reed.sorenson@utah.edu | Department of Biology | University of Utah" />

<meta name="date" content="2017-12-21" />

<title>RNAdecay Vignette 1: Decay Factor Normalization of RNA-seq RPM data for Decay Modeling</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore"><em>RNAdecay</em> Vignette 1: Decay Factor Normalization of RNA-seq RPM data for Decay Modeling</h1>
<h4 class="author"><em>Reed Sorenson | <a href="mailto:reed.sorenson@utah.edu">reed.sorenson@utah.edu</a> | Department of Biology | University of Utah</em></h4>
<h4 class="date"><em>2017-12-21</em></h4>



<p>The functions and code of the <code>RNAdecay</code> package were written for the purpose of analyzing decreasing levels of RNA abundance as measured by RNA short read sequencing (aka RNA-seq) after inhibition of transcription. Tissue is treated to block transcription (e.g., using a chemical such as cordycepin or ActinomycinD) and tissue is collected at regular intervals (referred to as the RNA decay time course, e.g., T0, T30, T60 refer to 0, 30, and 60 min of blocked transcription). Total RNA is extracted from the tissue and measured by RNA-seq. Although we wrote this package to analyze RNA-seq data, it could also be adapted to analyze individual gene data from quantitative Reverse Transcription Polymerase Chain Reaction (qRT-PCR) measures, however, normalization (as herein described) is best performed using a number of stable reference genes.</p>
<p>Raw (Illumina) RNA-seq data are typically libraries made of up 10M-250M short (50 nt) sequences. These sequences are aligned to the genome and counted based on the location of their alignment in annotated genomic ranges (e.g., genes, exons, etc). Read counts for each gene are normalized to the size of their respective RNA-seq library for comparison to libraries generated from other biological material, therefore, read counts are expressed as reads per million (RPM). We begin RNA double exponential decay analysis in this package with RPM RNA abundance data.</p>
<div id="t0-normalization" class="section level2">
<h2>1. T0 NORMALIZATION</h2>
<p>Prior to modeling RNA decay using the double exponential model, we normalize the read counts to the initial (T0) RNA abundance, the transcript level just prior to blocking transcription. This is unique for each gene and in an experiment with multiple replicates (reps), we use the mean value for all T0 reps. All RPM values in the time course are divided by the mean T0 value. For example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># you can see that in this fabricated data the RNAs have a half-life of ~30 min</span>
RPM_geneX =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">T0 =</span> <span class="kw">c</span>(<span class="dv">150</span>, <span class="dv">135</span>, <span class="dv">148</span>), <span class="dt">T30 =</span> <span class="kw">c</span>(<span class="dv">72</span>, <span class="dv">76</span>, <span class="dv">69</span>), <span class="dt">T60 =</span> <span class="kw">c</span>(<span class="dv">35</span>, <span class="dv">35</span>, <span class="dv">30</span>), 
                     <span class="dt">row.names =</span> <span class="kw">paste0</span>(<span class="st">&quot;rep&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>))
RPM_geneX</code></pre></div>
<pre><code>##       T0 T30 T60
## rep1 150  72  35
## rep2 135  76  35
## rep3 148  69  30</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">RPM_geneX_T0norm =<span class="st"> </span>RPM_geneX<span class="op">/</span><span class="kw">mean</span>(RPM_geneX<span class="op">$</span>T0)
RPM_geneX_T0norm</code></pre></div>
<pre><code>##             T0       T30       T60
## rep1 1.0392610 0.4988453 0.2424942
## rep2 0.9353349 0.5265589 0.2424942
## rep3 1.0254042 0.4780600 0.2078522</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colMeans</span>(RPM_geneX_T0norm)</code></pre></div>
<pre><code>##        T0       T30       T60 
## 1.0000000 0.5011547 0.2309469</code></pre>
<p>The following workflow simplifies this process on an example data set included in this package (from Sorenson et al., 2017) which we will use throughout the vignettes. It has 4 genotypes, 8 time points, and 4 biological replicates for 128 samples (columns) and 118 genes (rows). This script demends on column names consisting of unique, non-nested variable tags separated by underscores (’_’), rownames are gene ID’s.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RNAdecay)
RPMs[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">11</span>, <span class="dv">128</span>)] <span class="co"># built-in example data</span></code></pre></div>
<pre><code>##           WT_00_r1  WT_07_r1  WT_15_r1  WT_30_r1  WT_60_r1 WT_120_r1
## AT1G02860 10.35838  9.772902  9.682167  9.298616  8.478808  7.007951
## AT1G09660 19.77509 17.684298 19.717329 15.455808 15.164023 13.912843
##           WT_240_r1 WT_480_r1 sov_00_r1 sov_07_r1 sov_15_r1 vs_480_r4
## AT1G02860  5.482903  6.523851  11.78139  11.85849  9.238458  25.50327
## AT1G09660 10.342748 11.261410  21.48637  19.33097 18.476917  17.56892</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># </span><span class="al">NOTE</span><span class="co">: not showing all columns here</span></code></pre></div>
<p>Biological replicate data columns are indexed using the <code>cols()</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">cols</span>(RPMs, <span class="st">&quot;WT&quot;</span>, <span class="st">&quot;00&quot;</span>) <span class="co">#gives the column indices that contain both tags in the column names</span></code></pre></div>
<pre><code>## [1]  1 33 65 97</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colnames</span>(RPMs)[<span class="kw">cols</span>(RPMs, <span class="st">&quot;WT&quot;</span>, <span class="st">&quot;00&quot;</span>)]</code></pre></div>
<pre><code>## [1] &quot;WT_00_r1&quot; &quot;WT_00_r2&quot; &quot;WT_00_r3&quot; &quot;WT_00_r4&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># </span><span class="al">NOTE</span><span class="co">: this is based on grep pattern matching so tags MUST be unique and non-overlapping (i.e. one entire label should not be part of another).</span></code></pre></div>
<p>Replicate samples’ mean and standard error (SE) data.frames are generated by looping over label combinations and binding the values together.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create a directory for results output</span>
wrdir =<span class="st"> &quot;~/DecayAnalysis/Example analysis results 1&quot;</span>
<span class="cf">if</span>(<span class="op">!</span><span class="kw">file.exists</span>(wrdir)) <span class="kw">dir.create</span>(wrdir, <span class="dt">recursive =</span> <span class="ot">TRUE</span>)

<span class="co"># specify sample tags from colnames of RPMs</span>
treatment =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;WT&quot;</span>, <span class="st">&quot;sov&quot;</span>, <span class="st">&quot;vcs&quot;</span>, <span class="st">&quot;vs&quot;</span>) <span class="co"># </span><span class="al">NOTE</span><span class="co">: I did not use 'vcs sov' for the name of the double mutant, or cols(RPMs, &quot;vcs&quot;, &quot;00&quot;) would have indexed all T0 samples of both &quot;vcs&quot; and &quot;vcs&quot; sov&quot;; instead, I used 'vs'.</span>
reps =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;r1&quot;</span>, <span class="st">&quot;r2&quot;</span>, <span class="st">&quot;r3&quot;</span>, <span class="st">&quot;r4&quot;</span>)
tdecay =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;00&quot;</span>, <span class="st">&quot;07&quot;</span>, <span class="st">&quot;15&quot;</span>, <span class="st">&quot;30&quot;</span>, <span class="st">&quot;60&quot;</span>, <span class="st">&quot;120&quot;</span>, <span class="st">&quot;240&quot;</span>, <span class="st">&quot;480&quot;</span>) <span class="co">#again NO nesting of one label in another hence &quot;00&quot; instead of &quot;0&quot;.</span></code></pre></div>
<p>Loop over the treatments and timepoints to subset the appropriate columns, and calculate mean and SE of replicates.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mean_RPM =<span class="st"> </span><span class="kw">data.frame</span>(<span class="st">&quot;geneID&quot;</span> =<span class="st"> </span><span class="kw">rownames</span>(RPMs))
SE_RPM =<span class="st"> </span><span class="kw">data.frame</span>(<span class="st">&quot;geneID&quot;</span> =<span class="st"> </span><span class="kw">rownames</span>(RPMs))
<span class="cf">for</span>(g <span class="cf">in</span> treatment){
  <span class="cf">for</span>(t <span class="cf">in</span> tdecay){
    mean_RPM =<span class="st"> </span><span class="kw">cbind</span>(mean_RPM, <span class="kw">rowMeans</span>(RPMs[, <span class="kw">cols</span>(RPMs, g, t)]))
    <span class="kw">names</span>(mean_RPM)[<span class="kw">length</span>(mean_RPM)] =<span class="st"> </span><span class="kw">paste0</span>(g, <span class="st">&quot;_&quot;</span>, t)
    SE_RPM =<span class="st"> </span><span class="kw">cbind</span>(SE_RPM, <span class="kw">apply</span>(<span class="dt">X =</span> RPMs[, <span class="kw">cols</span>(RPMs, g, t)], <span class="dt">MARGIN =</span> <span class="dv">1</span>, <span class="dt">FUN =</span> stats<span class="op">::</span>sd)<span class="op">/</span><span class="kw">sqrt</span>(<span class="kw">length</span>(reps)))
    <span class="kw">names</span>(SE_RPM)[<span class="kw">length</span>(SE_RPM)] =<span class="st"> </span><span class="kw">paste0</span>(g, <span class="st">&quot;_&quot;</span>, t)
  }}
mean_RPM[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, ]</code></pre></div>
<pre><code>##              geneID    WT_00    WT_07    WT_15    WT_30     WT_60
## AT1G02860 AT1G02860 10.30576 10.92457 12.07227 10.18583  7.951206
## AT1G09660 AT1G09660 17.67649 17.60419 17.01889 15.56747 15.591995
##              WT_120    WT_240    WT_480  sov_00  sov_07   sov_15   sov_30
## AT1G02860  5.978875  4.907999  5.427231 10.4751 11.5972 10.76603 10.01565
## AT1G09660 13.271012 10.672795 10.035496 19.6877 19.3090 19.11892 18.04514
##              sov_60   sov_120   sov_240   sov_480   vcs_00   vcs_07
## AT1G02860  7.483476  5.882106  4.932124  5.242097 15.21011 16.66707
## AT1G09660 16.465382 13.815930 10.919155 11.141820 30.52888 28.87662
##             vcs_15   vcs_30   vcs_60   vcs_120   vcs_240   vcs_480
## AT1G02860 16.86892 14.45997 12.41270  9.471179  7.028314  6.223005
## AT1G09660 29.98092 26.93621 26.31798 24.230218 23.312002 22.954986
##              vs_00    vs_07    vs_15    vs_30    vs_60   vs_120   vs_240
## AT1G02860 23.98874 25.13802 24.56582 25.85643 23.09965 21.47141 20.11297
## AT1G09660 24.06467 23.25254 24.05339 21.89650 20.80315 22.72681 20.65078
##             vs_480
## AT1G02860 19.47309
## AT1G09660 20.31687</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">SE_RPM[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, ]</code></pre></div>
<pre><code>##              geneID     WT_00     WT_07     WT_15     WT_30     WT_60
## AT1G02860 AT1G02860 0.4482383 0.5495215 0.8677714 0.6957607 0.1826975
## AT1G09660 AT1G09660 1.0154376 0.9293058 0.9753225 0.9843020 1.3145629
##              WT_120    WT_240    WT_480    sov_00    sov_07    sov_15
## AT1G02860 0.4938045 0.2494622 0.5508578 0.4671639 0.5698232 0.6208221
## AT1G09660 0.7515276 0.6203130 0.8833117 0.7574344 0.4105105 0.5440400
##              sov_30    sov_60   sov_120   sov_240   sov_480   vcs_00
## AT1G02860 0.4900057 0.2521764 0.4775238 0.6288319 0.2047149 1.065802
## AT1G09660 0.9838571 0.8045619 0.7516455 0.4322530 0.7700577 1.149763
##              vcs_07    vcs_15    vcs_30    vcs_60   vcs_120   vcs_240
## AT1G02860 0.5303556 0.5052474 0.8726631 0.5491621 0.7702577 0.2626263
## AT1G09660 1.9180603 0.9563549 0.5693349 0.9244970 1.2275583 1.2626453
##             vcs_480    vs_00    vs_07    vs_15    vs_30    vs_60    vs_120
## AT1G02860 0.4689800 1.547137 1.006693 1.481889 2.374982 2.469157 2.7944529
## AT1G09660 0.9924902 2.215020 2.072957 1.388435 1.515420 1.204576 0.7517306
##              vs_240   vs_480
## AT1G02860 1.5576869 2.059173
## AT1G09660 0.4056269 1.758722</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># write output to file</span>
<span class="kw">write.table</span>(<span class="dt">x =</span> mean_RPM, <span class="kw">paste0</span>(wrdir, <span class="st">&quot;/RPM_mean.txt&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
<span class="kw">write.table</span>(<span class="dt">x =</span> SE_RPM,   <span class="kw">paste0</span>(wrdir, <span class="st">&quot;/RPM_SE.txt&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</code></pre></div>
<p>Optionally, an RPM-based filtering step could be applied at this point to remove lowly expressed genes, for example.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filt1 =<span class="st"> </span><span class="kw">rep</span>(<span class="ot">TRUE</span>, <span class="dv">118</span>) <span class="co"># we will not filter in this example, so the filter value for each gene is set to TRUE.</span></code></pre></div>
<p>Replicate samples’ mean T0 RPM values are then used to normalize gene level data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mT0norm =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">row.names =</span> <span class="kw">rownames</span>(RPMs)[filt1])
<span class="cf">for</span>(g <span class="cf">in</span> treatment){
  mean_T0reps =<span class="st"> </span><span class="kw">rowMeans</span>(RPMs[filt1, <span class="kw">cols</span>(RPMs, g, <span class="st">&quot;00&quot;</span>)])
  <span class="cf">for</span>(r <span class="cf">in</span> reps){
    df =<span class="st"> </span>RPMs[filt1, <span class="kw">colnames</span>(RPMs)[<span class="kw">cols</span>(RPMs, g, r)]]
    df =<span class="st"> </span>df[, <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(tdecay)]<span class="op">/</span>mean_T0reps
    mT0norm =<span class="st"> </span><span class="kw">cbind</span>(mT0norm, df)
  }}
<span class="kw">write.table</span>(<span class="dt">x =</span> mT0norm, <span class="dt">file =</span> <span class="kw">paste0</span>(wrdir, <span class="st">&quot;/mean T0 normalized.txt&quot;</span>),  <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</code></pre></div>
<p>Replicate samples’ mean and SE data.frames are then calculated on the T0 normalized data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mean_mT0norm =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">row.names =</span> <span class="kw">rownames</span>(mT0norm))
<span class="cf">for</span>(g <span class="cf">in</span> treatment){
  <span class="cf">for</span>(t <span class="cf">in</span> tdecay){
    mean_mT0norm =<span class="st"> </span><span class="kw">cbind</span>(mean_mT0norm, <span class="kw">rowMeans</span>(mT0norm[, <span class="kw">cols</span>(mT0norm, g, t)]))
    <span class="kw">names</span>(mean_mT0norm)[<span class="kw">length</span>(<span class="kw">names</span>(mean_mT0norm))] =<span class="st"> </span><span class="kw">paste0</span>(g, <span class="st">&quot;_&quot;</span>, t)
  }}

SE_mT0norm =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">row.names =</span> <span class="kw">rownames</span>(mT0norm))
<span class="cf">for</span>(g <span class="cf">in</span> treatment){
  <span class="cf">for</span>(t <span class="cf">in</span> tdecay){
    SE_mT0norm =<span class="st"> </span><span class="kw">cbind</span>(SE_mT0norm, <span class="kw">apply</span>(<span class="dt">X =</span> mT0norm[, <span class="kw">cols</span>(mT0norm, g, t)], <span class="dt">MARGIN =</span> <span class="dv">1</span>, <span class="dt">FUN =</span> <span class="cf">function</span>(x) stats<span class="op">::</span><span class="kw">sd</span>(x)<span class="op">/</span><span class="kw">sqrt</span>(<span class="kw">length</span>(reps))))
    <span class="kw">names</span>(SE_mT0norm)[<span class="kw">length</span>(<span class="kw">names</span>(SE_mT0norm))] =<span class="st"> </span><span class="kw">paste0</span>(g, <span class="st">&quot;_&quot;</span>, t)
  }}

<span class="co"># write output to file</span>
<span class="kw">write.table</span>(<span class="dt">x =</span> mean_mT0norm, <span class="dt">file =</span> <span class="kw">paste0</span>(wrdir, <span class="st">&quot;/mean T0 normalized_Mean.txt&quot;</span>),  <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
<span class="kw">write.table</span>(<span class="dt">x =</span> SE_mT0norm, <span class="dt">file =</span> <span class="kw">paste0</span>(wrdir, <span class="st">&quot;/mean T0 normalized_SE.txt&quot;</span>),  <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</code></pre></div>
</div>
<div id="decay-factor-correction" class="section level2">
<h2>2. DECAY FACTOR CORRECTION</h2>
<p>After inhibition of RNA synthesis,the total pool of RNA in cells decreases as degradation proceeds. As this occurs, very stable RNAs become a much larger proportion of the total RNA pool even though their concentration in the cell remains nearly constant. RNA-seq library normalization scales reads to library size (i.e. to the total RNA pool) and, because of this, abundance of stable RNAs appears to increase. The apparent increase in abundance of stable genes is proportional to the reduction in the total pool of RNA, so we can use it to apply a correction to the data. We call this decay factor correction. The decay factor is estimated based on the RPM increase (relative to the T0 samples) of a set of stable and abundant reference genes.</p>
<p>We selected 30 genes with high abundance on which to calculate the decay factors for each sample. These were then used to correct abundance in each of the respective samples. A unique decay factor is calculated for each treatment at each time point.</p>
<p>First, make a vector of ’geneID’s of abundant and stable reference genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">stablegenes =<span class="st"> </span><span class="kw">c</span>( <span class="st">&quot;ATCG00490&quot;</span>, <span class="st">&quot;ATCG00680&quot;</span>, <span class="st">&quot;ATMG00280&quot;</span>, <span class="st">&quot;ATCG00580&quot;</span>, <span class="st">&quot;ATCG00140&quot;</span>, <span class="st">&quot;AT4G38970&quot;</span>, <span class="st">&quot;AT2G07671&quot;</span>, <span class="st">&quot;ATCG00570&quot;</span>, <span class="st">&quot;ATMG00730&quot;</span>, <span class="st">&quot;AT2G07727&quot;</span>, <span class="st">&quot;AT2G07687&quot;</span>, <span class="st">&quot;ATMG00160&quot;</span> ,<span class="st">&quot;AT3G11630&quot;</span>, <span class="st">&quot;ATMG00060&quot;</span>, <span class="st">&quot;ATCG00600&quot;</span>, <span class="st">&quot;ATMG00220&quot;</span>, <span class="st">&quot;ATMG01170&quot;</span>, <span class="st">&quot;ATMG00410&quot;</span>, <span class="st">&quot;AT1G78900&quot;</span>, <span class="st">&quot;AT3G55440&quot;</span>, <span class="st">&quot;ATMG01320&quot;</span>, <span class="st">&quot;AT2G21170&quot;</span> ,<span class="st">&quot;AT5G08670&quot;</span>, <span class="st">&quot;AT5G53300&quot;</span>, <span class="st">&quot;ATMG00070&quot;</span>, <span class="st">&quot;AT1G26630&quot;</span>, <span class="st">&quot;AT5G48300&quot;</span>, <span class="st">&quot;AT2G33040&quot;</span>, <span class="st">&quot;AT5G08690&quot;</span>, <span class="st">&quot;AT1G57720&quot;</span>)</code></pre></div>
<p>Mean T0 normalized values of stable genes are then pulled out of the <code>mean_mT0norm</code> data.frame and used to calculate the decay factor for each set of replicate samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">stabletable =<span class="st"> </span>mean_mT0norm[stablegenes, ]
normFactors =<span class="st"> </span><span class="kw">colMeans</span>(stabletable)
<span class="kw">write.table</span>(<span class="dt">x =</span> normFactors, <span class="kw">paste0</span>(wrdir, <span class="st">&quot;/Normalziation Factors.txt&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
normFactors_mean =<span class="st"> </span><span class="kw">matrix</span>(normFactors, <span class="dt">nrow =</span> <span class="kw">length</span>(tdecay))
normFactors_SE =<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">apply</span>(<span class="dt">X =</span> stabletable, <span class="dt">MARGIN =</span> <span class="dv">2</span>, <span class="cf">function</span>(x) stats<span class="op">::</span><span class="kw">sd</span>(x)<span class="op">/</span><span class="kw">sqrt</span>(<span class="kw">length</span>(stablegenes))), <span class="dt">nrow =</span> <span class="kw">length</span>(tdecay))
t.decay =<span class="st"> </span><span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">7.5</span>, <span class="dv">15</span>, <span class="dv">30</span>, <span class="dv">60</span>, <span class="dv">120</span>, <span class="dv">240</span>, <span class="dv">480</span>)
<span class="kw">rownames</span>(normFactors_mean) =<span class="st"> </span>t.decay
<span class="kw">rownames</span>(normFactors_SE) =<span class="st"> </span>t.decay
<span class="kw">colnames</span>(normFactors_mean) =<span class="st"> </span>treatment
<span class="kw">colnames</span>(normFactors_SE) =<span class="st"> </span>treatment
<span class="kw">list</span>(<span class="dt">normalizationFactors =</span> normFactors_mean, <span class="dt">SE =</span> normFactors_SE)</code></pre></div>
<pre><code>## $normalizationFactors
##           WT      sov      vcs       vs
## 0   1.000000 1.000000 1.000000 1.000000
## 7.5 1.036541 1.018474 1.019401 1.006349
## 15  1.039264 1.049762 1.030781 1.016110
## 30  1.093192 1.089058 1.068392 1.057446
## 60  1.159184 1.150108 1.143107 1.127042
## 120 1.244396 1.254913 1.238043 1.211260
## 240 1.360045 1.361653 1.355177 1.331136
## 480 1.518916 1.524910 1.508710 1.490484
## 
## $SE
##               WT          sov          vcs           vs
## 0   3.764009e-18 9.219900e-18 8.416579e-18 5.323112e-18
## 7.5 7.187629e-03 6.156141e-03 6.354406e-03 3.671789e-03
## 15  8.350257e-03 8.745853e-03 4.268441e-03 7.639774e-03
## 30  1.018092e-02 9.213515e-03 8.169738e-03 9.683929e-03
## 60  8.408549e-03 9.406738e-03 7.099825e-03 1.257889e-02
## 120 1.081113e-02 1.397098e-02 1.073154e-02 1.346151e-02
## 240 1.827014e-02 1.776487e-02 1.474298e-02 1.564277e-02
## 480 3.641681e-02 4.239630e-02 3.448004e-02 3.649816e-02</code></pre>
<p>Generate a matrix of the same dimensions as <code>mT0norm</code> with the appropriate decay factors in the same respective positions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nF =<span class="st"> </span><span class="kw">vector</span>()
ind =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">names</span>(normFactors), <span class="cf">function</span>(x) <span class="kw">grep</span>(x, <span class="kw">colnames</span>(mT0norm)))
<span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(ind)){
nF[ind[, i]] =<span class="st"> </span><span class="kw">colnames</span>(ind)[i]
}
normFactorsM =<span class="st"> </span><span class="kw">t</span>(<span class="kw">matrix</span>(<span class="kw">rep</span>(normFactors[nF], <span class="kw">nrow</span>(mT0norm)), <span class="dt">ncol =</span> <span class="kw">nrow</span>(mT0norm)))
<span class="kw">rm</span>(nF, ind)</code></pre></div>
<p>Apply the decay factor corrections.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mT0norm_<span class="dv">2</span> =<span class="st"> </span><span class="kw">data.frame</span>(mT0norm<span class="op">/</span>normFactorsM, 
                     <span class="dt">row.names =</span> <span class="kw">rownames</span>(mT0norm))

<span class="kw">write.table</span>(mT0norm_<span class="dv">2</span>, <span class="kw">paste0</span>(wrdir, <span class="st">&quot;/mean T0 normalized and decay factor corrected.txt&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</code></pre></div>
<p>Rearrange the RNA decay data into a long form data.frame for modeling and visualization.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mT0norm_<span class="fl">2.1</span> =<span class="st"> </span>reshape2<span class="op">::</span><span class="kw">melt</span>(<span class="kw">as.matrix</span>(mT0norm_<span class="dv">2</span>), <span class="dt">varnames =</span> <span class="kw">c</span>(<span class="st">&quot;geneID&quot;</span>, <span class="st">&quot;variable&quot;</span>))
mT0norm_<span class="fl">2.1</span> =<span class="st"> </span><span class="kw">cbind</span>(mT0norm_<span class="fl">2.1</span>, reshape2<span class="op">::</span><span class="kw">colsplit</span>(mT0norm_<span class="fl">2.1</span><span class="op">$</span>variable, <span class="st">&quot;_&quot;</span>, <span class="dt">names =</span> <span class="kw">c</span>(<span class="st">&quot;treatment&quot;</span>, <span class="st">&quot;t.decay&quot;</span>, <span class="st">&quot;rep&quot;</span>)))

mT0norm_<span class="fl">2.1</span> =<span class="st"> </span>mT0norm_<span class="fl">2.1</span>[, <span class="kw">colnames</span>(mT0norm_<span class="fl">2.1</span>) <span class="op">!=</span><span class="st">  &quot;variable&quot;</span>]
mT0norm_<span class="fl">2.1</span> =<span class="st"> </span>mT0norm_<span class="fl">2.1</span>[, <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">2</span>)]
<span class="kw">colnames</span>(mT0norm_<span class="fl">2.1</span>) =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;geneID&quot;</span>, <span class="st">&quot;treatment&quot;</span>, <span class="st">&quot;t.decay&quot;</span>, <span class="st">&quot;rep&quot;</span>, <span class="st">&quot;value&quot;</span>)
mT0norm_<span class="fl">2.1</span><span class="op">$</span>rep =<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;r&quot;</span>, <span class="st">&quot;rep&quot;</span>, mT0norm_<span class="fl">2.1</span><span class="op">$</span>rep)
mT0norm_<span class="fl">2.1</span><span class="op">$</span>t.decay =<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">gsub</span>(<span class="st">&quot;7&quot;</span>, <span class="st">&quot;7.5&quot;</span>, <span class="kw">as.numeric</span>(mT0norm_<span class="fl">2.1</span><span class="op">$</span>t.decay)))
mT0norm_<span class="fl">2.1</span><span class="op">$</span>treatment =<span class="st"> </span><span class="kw">factor</span>(mT0norm_<span class="fl">2.1</span><span class="op">$</span>treatment,<span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;WT&quot;</span>,<span class="st">&quot;sov&quot;</span>,<span class="st">&quot;vcs&quot;</span>,<span class="st">&quot;vs&quot;</span>))
mT0norm_<span class="fl">2.1</span><span class="op">$</span>rep =<span class="st"> </span><span class="kw">factor</span>(mT0norm_<span class="fl">2.1</span><span class="op">$</span>rep, <span class="dt">levels =</span> <span class="kw">paste0</span>(<span class="st">&quot;rep&quot;</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>))
<span class="kw">write.table</span>(<span class="dt">x =</span> mT0norm_<span class="fl">2.1</span>,  <span class="dt">file =</span> <span class="kw">paste0</span>(wrdir, 
<span class="st">&quot;/ExampleDecayData+stableGenes.txt&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</code></pre></div>
<p>Following the steps in this vignettes, the resulting normalized data frame should be identical to the one supplied in this package called decaydata. To check:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(decaydata[,<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>] <span class="op">==</span><span class="st"> </span>mT0norm_<span class="fl">2.1</span>[,<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>]) <span class="co"># should be all TRUE no FALSE</span></code></pre></div>
<pre><code>## 
##  TRUE 
## 60416</code></pre>
<p>The normalized data is now ready to be used for modeling decay rates described in <em>RNAdecay Vignette 2: Double Exponential RNA Decay Modeling</em>.</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
